File: packages/mikan/src/title.ts
export interface ParsedTitle {
  episode: number | null
  season: number | null
  title: string
}

const QUALITY_NUMBERS = new Set([360, 480, 720, 1080, 1440, 2160, 4320])

export function parseMikanTitle(raw: string): ParsedTitle {
  const text = raw.trim()
  if (!text) {
    return { title: '', season: null, episode: null }
  }

  let working = text
  let season: number | null = null
  let episode: number | null = null

  const se = working.match(/s(\d{1,2})e(\d{1,4})/i)
  if (se) {
    season = Number(se[1])
    episode = Number(se[2])
    working = working.replace(se[0], ' ')
  }

  if (season === null) {
    const numbered = working.match(/第(\d{1,2})季/)
    const named = working.match(/season\s*(\d{1,2})/i)
    const hit = numbered ?? named
    if (hit) {
      season = Number(hit[1])
      working = working.replace(hit[0], ' ')
    }
  }

  if (episode === null) {
    const ji = working.match(/第\s*(\d{1,4})\s*[期話话集]/)
    if (ji) {
      episode = Number(ji[1])
      working = working.replace(ji[0], ' ')
    }
  }

  if (episode === null) {
    const dash = working.match(/(?:^|\D)-\s*(\d{1,4})(?=\s*(?:\[|$|v\d))/i)
    if (dash) {
      episode = Number(dash[1])
      working = working.replace(new RegExp(`-\\s*${dash[1]}`), ' ')
    }
  }

  if (episode === null) {
    for (const tag of working.matchAll(/\[(\d{1,4})(?:v\d+)?]/gi)) {
      const value = Number(tag[1])
      if (!QUALITY_NUMBERS.has(value) && value < 1900) {
        episode = value
        working = working.replace(tag[0], ' ')
        break
      }
    }
  }

  let title = working.replaceAll(/\[[^\]]*]/g, ' ')
  title = title.replaceAll(/第\s*\d+\s*[季期話话集]/g, ' ')
  title = title.replaceAll(/(?:^|\D)-\s*\d+\b/g, ' ')
  title = title.replaceAll(/\s+/g, ' ').trim()

  if (title.includes('/')) {
    const parts = title
      .split('/')
      .map((part) => part.trim())
      .filter(Boolean)
    const cjk = parts.find((part) => /[\u3400-\u9FFF]/.test(part))
    title = (cjk ?? parts.at(-1) ?? title).trim()
  }

  title = title.replaceAll(/^[/:–—：-]+|[/:–—：-]+$/g, '').trim()

  return { title, season, episode }
}

File: apps/helper-go/internal/mikan/title.go
package mikan

import (
	"regexp"
	"strconv"
	"strings"
)

type ParsedTitle struct {
	Title   string
	Season  *int
	Episode *int
}

var qualityNumbers = map[int]struct{}{
	360: {}, 480: {}, 720: {}, 1080: {}, 1440: {}, 2160: {}, 4320: {},
}

var (
	reSE          = regexp.MustCompile(`(?i)S(\d{1,2})E(\d{1,4})`)
	reSeasonZh    = regexp.MustCompile(`第(\d{1,2})季`)
	reSeasonEn    = regexp.MustCompile(`(?i)Season\s*(\d{1,2})`)
	reJi          = regexp.MustCompile(`第\s*(\d{1,4})\s*[集话話期]`)
	reDash        = regexp.MustCompile(`(?i)(?:^|[^0-9])-\s*(\d{1,4})`)
	reDashFollow  = regexp.MustCompile(`(?i)^\s*(\[|v\d|$)`)
	reTagNum      = regexp.MustCompile(`(?i)\[(\d{1,4})(?:v\d+)?\]`)
	reBrackets    = regexp.MustCompile(`\[[^\]]*\]`)
	reOrdinal     = regexp.MustCompile(`第\s*\d+\s*[集话話期季]`)
	reDashNumber  = regexp.MustCompile(`(?:^|\D)-\s*\d+\b`)
	reEdgePunct   = regexp.MustCompile(`^[-–—/:：]+|[-–—/:：]+$`)
	reCJK         = regexp.MustCompile(`[\x{3400}-\x{9FFF}]`)
)

func ParseMikanTitle(raw string) ParsedTitle {
	text := strings.TrimSpace(raw)
	if text == "" {
		return ParsedTitle{}
	}

	working := text
	var season, episode *int

	if se := reSE.FindStringSubmatch(working); se != nil {
		s := atoi(se[1])
		e := atoi(se[2])
		season = &s
		episode = &e
		working = strings.Replace(working, se[0], " ", 1)
	}

	if season == nil {
		hit := reSeasonZh.FindStringSubmatch(working)
		if hit == nil {
			hit = reSeasonEn.FindStringSubmatch(working)
		}
		if hit != nil {
			s := atoi(hit[1])
			season = &s
			working = strings.Replace(working, hit[0], " ", 1)
		}
	}

	if episode == nil {
		if ji := reJi.FindStringSubmatch(working); ji != nil {
			e := atoi(ji[1])
			episode = &e
			working = strings.Replace(working, ji[0], " ", 1)
		}
	}

	if episode == nil {
		if loc := reDash.FindStringSubmatchIndex(working); loc != nil {
			rest := working[loc[1]:]
			if reDashFollow.MatchString(rest) {
				e := atoi(working[loc[2]:loc[3]])
				episode = &e
				re := regexp.MustCompile(`-\s*` + regexp.QuoteMeta(working[loc[2]:loc[3]]))
				working = re.ReplaceAllString(working, " ")
			}
		}
	}

	if episode == nil {
		for _, tag := range reTagNum.FindAllStringSubmatch(working, -1) {
			value := atoi(tag[1])
			if _, quality := qualityNumbers[value]; !quality && value < 1900 {
				episode = &value
				working = strings.Replace(working, tag[0], " ", 1)
				break
			}
		}
	}

	title := reBrackets.ReplaceAllString(working, " ")
	title = reOrdinal.ReplaceAllString(title, " ")
	title = reDashNumber.ReplaceAllString(title, " ")
	title = strings.Join(strings.Fields(title), " ")
	title = strings.TrimSpace(title)

	if strings.Contains(title, "/") {
		parts := strings.Split(title, "/")
		var kept []string
		for _, part := range parts {
			part = strings.TrimSpace(part)
			if part != "" {
				kept = append(kept, part)
			}
		}
		chosen := title
		for _, part := range kept {
			if reCJK.MatchString(part) {
				chosen = part
				break
			}
		}
		if chosen == title && len(kept) > 0 {
			chosen = kept[len(kept)-1]
		}
		title = strings.TrimSpace(chosen)
	}

	title = reEdgePunct.ReplaceAllString(title, "")
	title = strings.TrimSpace(title)
	return ParsedTitle{Title: title, Season: season, Episode: episode}
}

func atoi(raw string) int {
	n, _ := strconv.Atoi(raw)
	return n
}
